gitbootcamp.rst

```diff
@@ -0,0 +1,244 @@
+them to the remote repository.
+
+::
+
+   $ git checkout some-branch
+   $ git push origin some-branch
+
+
+Creating a Pull Request
+-----------------------
+
+1. Go to https://github.com/python/cpython.
+
+2. Click ``compare across forks`` link.
+
+3. Select the base fork: ``python/cpython`` and base branch: ``master``
+
+4. Select the head fork: ``UserName/cpython`` and base branch: the branch
+   containing your changes.
+
+5. Press ``Create Pull Request`` button.
+
+
+Syncing With Upstream
+---------------------
+
+Scenario:
+
+- You forked cpython repository some time ago.
+- Time passes.
+- There have been new commits made in upstream cpython repository.
+- Your forked cpython repository is no longer up to date.
+- You now want to update your forked cpython repository to be the same as
+  upstream.
+
+Solution::
+
+   $ git checkout master
+   $ git fetch upstream
+   $ git rebase upstream/master
+   $ git push origin master
+
+
+Another scenario:
+
+- You created ``some-branch`` some time ago.
+- Time passes.
+- You made some commits to ``some-branch``.
+- Meanwhile, there are recent changes from upstream cpython repository.
+- You want to incorporate the recent changes from upstream into ``some-branch``.
+
+Solution::
+
+   $ git checkout some-branch
+   $ git fetch upstream
+   $ git rebase upstream/master
+
+
+Backporting Merged Changes
+--------------------------
+
+When a pull request has been merged to master, and it needs to be backported
+into one of the maintenance branches.
+
+First, obtain the commit sha1 from the merged pull request:
+
+1. Go to the merged pull request page, for example::
+
+     https://github.com/python/cpython/pull/PR-ID
+
+2. Scroll down and find the activity that says something like::
+
+     SomeUser merged commit <hash> into python:master ...
+
+3. Follow the link to <hash>.
+
+4. Copy the complete hash value.
+
+The commit hash will be used below.
+
+To backport the commit to 3.6::
+
+   $ git fetch upstream
+   $ git checkout -b backport-someissue-3.6 upstream/3.6
+   $ git cherry-pick -x hashvalue
+   $ git push origin backport-someissue-3.6
+
+Go to https://github.com/python/cpython to create the pull request.  Select
+``3.6`` as the base branch, and ``backport-someissue-3.6`` as the head branch.
+
+
+Downloading Other's Patches
+---------------------------
+
+Scenario:
+
+- UserX made a pull request to cpython.
+- Before merging it, you want to be able to test their changes locally.
+
+Solution:
+
+1. Go to the pull request page, e.g.::
+
+      https://github.com/python/cpython/pull/PR-ID
+
+2. Under the pull request description, you will see something like::
+
+      UserX wants to merge 1 commit into python:master from UserX:fixspamlib
+
+3. You'll need to first clone their cpython repo, and then checkout their
+   branch::
+
+      $ mkdir UserX
+      $ cd UserX
+      $ git clone https://github.com/UserX/cpython
+      $ cd cpython
+      $ git checkout fixspamlib
```

Comment:
This can be simplified significantly.  GitHub offers pull requests as remote branches that are not pulled automatically, but you can pull them explicitly by doing `git fetch upstream pull/${1}/head:pr/${1} && git checkout pr/${1}` where `${1}` is the PR number.  @berkerpeksag has this nicely packaged up as an alias:

```
[alias]
        pr = !sh -c \"git fetch upstream pull/${1}/head:pr/${1} && git checkout pr/${1}\" -
```

which allows one to do `git pr 615` to fetch and check out python/cpython#615.  The local branch name (`pr/${1}` or `pr/615` here) can be adjusted as you see fit; I might prefer `pr_${1}`.